event/google.go

```
@@ -0,0 +1,142 @@
+// We need a specific Google publisher because ordering doesn't generalize well.
+// All ordered publishers should implement the same interface.
+// Call [OrderedGooglePublisher.Shutdown] to stop all goroutines/clean up all resources.
+//
+// Region is required since it is a best practice to publish all messages within the same region:
+//   - https://cloud.google.com/pubsub/docs/publish-best-practices#ordering
+//
+// It must be a valid Google cloud region, it is used to defined the publish endpoint.
+func NewOrderedGooglePublisher[T any](ctx context.Context, project, region, topicName, eventName string) (*OrderedGooglePublisher[T], error) {
+	client, err := pubsub.NewClient(ctx, project, option.WithEndpoint(region+"-pubsub.googleapis.com:443"))
+	if err != nil {
+		return nil, fmt.Errorf("creating pubsub client: %w", err)
+	}
+	topic := client.Topic(topicName)
+	topic.EnableMessageOrdering = true
+	return &OrderedGooglePublisher[T]{eventName: eventName, client: client, topic: topic}, nil
+}
+
+// Publish will publish the given event with the given ordering key.
+// If an unrecoverable error happens an [ErrUnrecoverable] will be returned, when that happens if
+// [OrderedGooglePublisher.Resume] is not called with the same orderingKey all subsequent
+// [Publish] calls with that orderingKey will fail.
+// This allows clients to control the ordering behavior when something went wrong, resuming will
+// discard the failed publish and will result in an out of order stream.
+func (p *OrderedGooglePublisher[T]) Publish(ctx context.Context, event T, orderingKey string) error {
+	encBody, err := serializeEvent(ctx, p.eventName, event)
+	if err != nil {
+		return err
+	}
+
+	sample := publishSampler()
+	res := p.topic.Publish(ctx, &pubsub.Message{
+		OrderingKey: orderingKey,
+		Data:        encBody,
+	})
+	_, err = res.Get(ctx)
+	sample(p.eventName, len(encBody), err)
+
+	return xerrors.Tag(err, ErrUnrecoverable)
```

noticed that you allowed calling it with `nil` but I wonder if this is a good practice.
Usually, the idea of hiding error handling does not Go well.

---

Well...errors.Join on the stdlib does this =P...not sure it is a terrible idea...but no strong opinions...but then I would add an immediate panic there to avoid failing later in odd ways.

---

yeah I see some cases where it can be useful, coalescing non-nil values is a good one but not sure in general.
In this case it's a bit odd, because if supported I imagine people will do this often and then it's a new trend of bubbling up errors with optional wrappers without err handling =)